micrometer-core/src/main/java/io/micrometer/core/instrument/observation/ObservationOrTimerCompatibleInstrumentation.java

```diff
@@ -0,0 +1,141 @@
+    private Timer.Sample sample;
+
+    @Nullable
+    private Observation observation;
+
+    @Nullable
+    private T context;
+
+    /**
+     * Start timing based on Observation and the convention for it if
+     * {@link ObservationRegistry} is not null and not the no-op registry. Otherwise,
+     * timing will be instrumented with a {@link Timer} using the {@link MeterRegistry}.
+     * @param meterRegistry registry for Timer-based instrumentation
+     * @param observationRegistry registry for Observation-based instrumentation
+     * @param context supplier for the context to use if instrumenting with Observation
+     * @param convention convention that overrides the default convention and any
+     * conventions configured on the registry, if not null
+     * @param defaultConvention convention to use if one is not configured
+     * @return a started instrumentation
+     * @param <T> context type if Observation used for instrumentation
+     */
+    public static <T extends Observation.Context> ObservationOrTimerCompatibleInstrumentation<T> start(
+            MeterRegistry meterRegistry, @Nullable ObservationRegistry observationRegistry, Supplier<T> context,
+            @Nullable Observation.ObservationConvention<T> convention,
+            Observation.ObservationConvention<T> defaultConvention) {
+        ObservationOrTimerCompatibleInstrumentation<T> observationOrTimer = new ObservationOrTimerCompatibleInstrumentation<>(
+                meterRegistry, observationRegistry, convention, defaultConvention);
+        observationOrTimer.start(context);
+        return observationOrTimer;
+    }
+
+    private ObservationOrTimerCompatibleInstrumentation(MeterRegistry meterRegistry,
+            @Nullable ObservationRegistry observationRegistry,
+            @Nullable Observation.ObservationConvention<T> convention,
+            Observation.ObservationConvention<T> defaultConvention) {
+        this.meterRegistry = meterRegistry;
+        this.observationRegistry = observationRegistry == null ? ObservationRegistry.NOOP : observationRegistry;
+        this.convention = convention;
+        this.defaultConvention = defaultConvention;
+    }
+
+    private void start(Supplier<T> contextSupplier) {
+        if (observationRegistry.isNoop()) {
+            sample = Timer.start(meterRegistry);
+        }
+        else {
+            this.context = contextSupplier.get();
+            observation = Observation.start(convention, defaultConvention, context, observationRegistry);
+        }
+    }
+
+    /**
+     * If using an Observation for instrumentation and the context is a
+     * {@link RequestReplySenderContext}, set the response object on it. Otherwise, do
+     * nothing.
+     * @param response response for the RequestReplySenderContext
+     * @param <RES> type of the response
+     */
+    public <RES> void setResponse(RES response) {
+        if (observationRegistry.isNoop() || !(context instanceof RequestReplySenderContext)) {
```

Comment:
What about `RequestReplyReceiverContext` ? We have the response there too?

Reply:
Good catch. I had only tried this with the Apache HTTP client instrumentation, which is why I probably only thought of this. I expect we'll catch other things and add to it as we use this in more places.
Do you think it makes sense to extract the response-related methods to an interface so we could check for that here and just cast to that?

Reply:
I'll follow-up with this in a separate pull request. I've left a TODO comment for now.